Add -fentry-point to glslc

dneto0 opened the pull request:
Add test mixin classes to check for substring in assembly
output.

Pass the path to the spirv-dis to the tests, and store
it in the test manager class.

dneto0 commented:
I can see how the -fentry-point tests should be moved to their own .py file.  Let me know if you want that.

antiagainst commented on glslc/test/expect.py:
```diff
@@ -212,6 +213,31 @@ def check_object_file_preamble(self, status):
+                stderr=subprocess.PIPE, cwd=status.directory)
+            output = process.communicate(None)
+            disassembly = output[0]
+        if self.expected_assembly_substr not in disassembly :
```
Extra space before colon.

antiagainst commented on glslc/test/glslc_test_framework.py:
```diff
@@ -198,8 +199,9 @@ def actual_decorator(cls):
     """Manages and runs a set of tests."""
 
-    def __init__(self, executable_path):
+    def __init__(self, executable_path, disassembler_path):
```
Maybe we should change `executable_path` to `glslc_path`?

antiagainst commented on glslc/test/option_dash_x.py:
```diff
@@ -51,11 +51,41 @@ class TestDashXGlslOnHlslShader(expect.ErrorMessageSubstr):
+
+@inside_glslc_testsuite('OptionDashX')
+class TestDashXHlslDashFEntryPointAffectsAllLaterFiles(expect.ValidObjectFileWithAssemblySubstr):
+    """Tests -x hlsl on an HLSL shader with -fentry-point."""
```
Comments need to be updated.

antiagainst commented on glslc/test/option_dash_x.py:
```diff
@@ -51,11 +51,41 @@ class TestDashXGlslOnHlslShader(expect.ErrorMessageSubstr):
 
     shader = FileShader(HLSL_VERTEX_SHADER, '.vert')
-    glslc_args = ['-x', 'hlsl', '-c', shader]
+    glslc_args = ['-x', 'hlsl', '-fentry-point=EntryPoint', '-S', '-c', shader]
```
redundant `-c`

antiagainst commented on glslc/test/option_dash_x.py:
```diff
@@ -51,11 +51,41 @@ class TestDashXGlslOnHlslShader(expect.ErrorMessageSubstr):
+    glslc_args = ['-x', 'hlsl', '-fentry-point=foobar', '-fentry-point=EntryPoint',
+                  '-c', shader]
+    expected_assembly_substr = "OpEntryPoint Vertex %EntryPoint \"EntryPoint\""
 
```
Test cases for:
* `-x hlsl -fentry-point=foo shader`
* `shader1 -fentry-point=foo shader2 -fentry-point=bar shader3`
?

dneto0 commented:
Addressed all feedback.

Also tweaked the FileShader class in a non-trivial way.  Test expectations for assembly substring is now on a per-shader basis, which was required for your requested test.

antiagainst commented on glslc/test/option_fentry_point.py:
```diff
@@ -0,0 +1,109 @@
+# Copyright 2015 The Shaderc Authors. All rights reserved.
```
2016

antiagainst commented on glslc/test/option_fentry_point.py:
```diff
@@ -0,0 +1,109 @@
+    glslc_args = ['-x', 'hlsl', '-fentry-point=EntryPoint', '-c', shader1, shader2]
+
+
+@inside_glslc_testsuite('OptionDashFEntryPoint')
```
I'd suggest to use the same test suite name so that they are grouped together.

antiagainst commented on glslc/test/expect.py:
```diff
@@ -280,6 +308,27 @@ def check_assembly_file_preamble(self, status):
+class ValidAssemblyFileWithSubstr(ValidAssemblyFile):
+    """Mixin class for checking that every input file generates a valid assembly
+    file following the assembly file naming rule, there is no output on
+    stdout/stderr, and the assembly file has a given substring."""
```
```
... and all assembly files have the given substring specified by expected_assembly_substr of the subclasses.

To mix in this class, subclasses need to provide expected_assembly_substr as the expected substring.
```

antiagainst commented on glslc/test/option_fentry_point.py:
```diff
@@ -0,0 +1,109 @@
+
+    shader = FileShader(GLSL_VERTEX_SHADER, '.vert')
+    glslc_args = ['-S', shader]
+    expected_assembly_substr = "OpEntryPoint Vertex %main \"main\""
```
ASSEMBLY_MAIN?

Applies for following cases.

dneto0 commented:
Addressed all feedback.

dneto0 commented:
Rebased, squashed and pushed into master as  89eefb94d
